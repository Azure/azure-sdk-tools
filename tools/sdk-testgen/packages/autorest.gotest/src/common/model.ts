import { ExampleModel, MockTestDefinitionModel, TestDefinitionModel } from '@autorest/testmodeler/dist/src/core/model';

interface GoFileData {
    packageName: string;
    packagePath?: string;
    imports: string;
}

export class GoMockTestDefinitionModel extends MockTestDefinitionModel implements GoFileData {
    packageName: string;
    imports: string;
}

export type GoTestDefinition = TestDefinitionModel & GoFileData;

export class GoExampleModel extends ExampleModel {
    opName: string;
    isLRO: boolean;
    isPageable: boolean;
    isMultiRespOperation: boolean;
    methodParametersOutput: string;
    methodParametersPlaceholderOutput: string;
    clientParametersOutput: string;
    clientParametersPlaceholderOutput: string;
    returnInfo: string[];
    checkResponse: boolean;
    responseOutput: string;
    responseType: string;
    responseIsDiscriminator: boolean;
    responseTypePointer: boolean;
    pollerType: string;
    pageableItemName: string;
}
