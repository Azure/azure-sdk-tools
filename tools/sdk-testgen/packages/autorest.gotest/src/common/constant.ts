export enum Config {
    exportCodemodel = 'testmodeler.export-codemodel',
    generateMockTest = 'testmodeler.generate-mock-test',
    generateSdkExample = 'testmodeler.generate-sdk-example',
    generateScenarioTest = 'testmodeler.generate-scenario-test',
    parents = '__parents',
    outputFolder = 'output-folder',
    module = 'module',
    filePrefix = 'file-prefix',
}

export enum TargetMode {
    mockTest = 'MOCK-TEST',
    scenarioTest = 'SCENARIO-TEST',
    sample = 'SAMPLE',
}
