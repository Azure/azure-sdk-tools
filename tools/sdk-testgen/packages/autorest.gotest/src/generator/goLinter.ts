import * as path from 'path';
import { AutorestExtensionHost } from '@autorest/extension-base';
import { Config, configDefaults } from '../common/constant';
import { Helper } from '@autorest/testmodeler/dist/src/util/helper';
import { TestConfig } from '@autorest/testmodeler/dist/src/common/testConfig';

export async function processRequest(host: AutorestExtensionHost): Promise<void> {
    const testConfig = new TestConfig(await host.GetValue(''), configDefaults);
    const files = await host.listInputs();
    Helper.execSync(`go install golang.org/x/tools/cmd/goimports@latest`);
    for (const outputFile of files) {
        if (outputFile.endsWith('.go')) {
            const pathName = path.join(testConfig.getValue(Config.outputFolder), outputFile);
            Helper.execSync(`go fmt ${pathName}`);
            Helper.execSync(`goimports -w ${pathName}`);
        }
    }
}
