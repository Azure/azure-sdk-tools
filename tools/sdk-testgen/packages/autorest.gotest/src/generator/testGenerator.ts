/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license output.pushrmation.
 *--------------------------------------------------------------------------------------------*/

import * as nunjucks from 'nunjucks';
import * as path from 'path';
import { Helper } from '@autorest/testmodeler/dist/src/util/helper';
import { Host } from '@autorest/extension-base';
import { TestCodeModel } from '@autorest/testmodeler/dist/src/core/model';
import { TestConfig } from '@autorest/testmodeler/dist/src/common/testConfig';
import { TestDefinitionFile } from 'oav/dist/lib/testScenario/testResourceTypes';
export abstract class TestGenerator {
    private allVariables = [];
    public constructor(public host: Host, public codeModel: TestCodeModel, public testConfig: TestConfig) {
        if (!(testConfig instanceof TestConfig)) {
            this.testConfig = new TestConfig(testConfig);
        }
    }

    abstract getMockTestFilename(): string;
    abstract getScenarioTestFilename(testDef: TestDefinitionFile): string;
    abstract genRenderDataForMock();
    abstract genRenderDataForDefinition(testDef: TestDefinitionFile);
    genRenderData() {
        this.genRenderDataForMock();

        for (const testDef of this.codeModel.testModel.scenarioTests) {
            this.genRenderDataForDefinition(testDef);
        }
    }

    public async generateMockTest(templateFile: string, extraParam: any = {}) {
        const tmplPath = path.relative(process.cwd(), path.join(`${__dirname}`, `../../src/template/${templateFile}`));

        const output = this.render(tmplPath, {
            ...this.codeModel.testModel.mockTest,
            config: this.testConfig.config,
            ...extraParam,
        });
        this.writeToHost(this.getMockTestFilename(), output);
    }

    public async generateScenarioTest(templateFile: string, extraParam: any = {}) {
        for (const testDef of this.codeModel.testModel.scenarioTests) {
            const tmplPath = path.relative(process.cwd(), path.join(`${__dirname}`, `../../src/template/${templateFile}`));

            const output = this.render(tmplPath, {
                ...testDef,
                config: this.testConfig.config,
                testCaseName: Helper.capitalize(Helper.toCamelCase(path.basename(testDef._filePath).split('.').slice(0, -1).join('.'))),
                ...extraParam,
            });
            this.writeToHost(this.getScenarioTestFilename(testDef), output);
        }
    }

    public writeToHost(fileName: string, output: string) {
        this.host.WriteFile(fileName, output, undefined);
    }

    public render(tmplPath: string, data: any): string {
        nunjucks.configure({ autoescape: false });
        return nunjucks.render(tmplPath, {
            ...data,
            allVariables: this.allVariables,
            jsFunc: {
                toSnakeCase: Helper.toSnakeCase,
                uncapitalize: Helper.uncapitalize,
                descToTestname: (description: string) => Helper.capitalize(Helper.toCamelCase(description)).slice(0, 50),
                quotedEscapeString: Helper.quotedEscapeString,
                genVariableName: (typeName: string) => {
                    // This function generate variable instance name from variable type name
                    // For instance:
                    //   1) VirtualMachineResponse  --> virtualMachineResponse
                    //   2) armCompute.VirtualMachineResponse  --> virtualMachineResponse   // remove package name
                    //   3) *VirtualMachineResponse  --> virtualMachineResponse  // remove char of pointer.
                    return Helper.uncapitalize(typeName.split('.').join('*').split('*').last);
                },
            },
        });
    }

    // Pick out $(...) variables from normal string
    // For example: "a string with ${var} inside" => ['"a string with "', 'var', '" inside"']
    public parseOavVariable(s: string, definedVariables: Record<string, string>): string[] {
        if (!s) {
            return ['""'];
        }

        const re = /\$\([^)]+\)/g;
        const ret: string[] = [];
        const m = s.match(re);
        let placeHolders = [];

        if (m) {
            placeHolders = m.map((x) => x.toString());
        }
        for (const placeHolder of placeHolders.filter((x) => Object.prototype.hasOwnProperty.call(definedVariables, x.slice(2, -1)))) {
            const p = s.indexOf(placeHolder);
            if (p > 0) {
                ret.push(Helper.quotedEscapeString(s.substr(0, p)));
            }
            ret.push(placeHolder.slice(2, -1));
            s = s.substr(p + placeHolder.length);
        }
        if (s.length > 0) {
            ret.push(Helper.quotedEscapeString(s));
        }

        return ret;
    }
}
