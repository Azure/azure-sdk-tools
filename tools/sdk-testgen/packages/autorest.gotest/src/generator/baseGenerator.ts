/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license output.pushrmation.
 *--------------------------------------------------------------------------------------------*/

import * as nunjucks from 'nunjucks';
import * as path from 'path';
import { GenerateContext } from './generateContext';
export abstract class BaseDataRender {
    public constructor(public context: GenerateContext) {}

    abstract renderData(): void;
}

export abstract class BaseCodeGenerator {
    public constructor(public context: GenerateContext) {}

    abstract generateCode(extraParam: Record<string, unknown>): void;

    protected renderAndWrite(model: any, templateFileName: string, outputFileName: string, extraParam: Record<string, unknown> = {}, jsFunc: Record<string, unknown> = {}) {
        const tmplPath = path.relative(process.cwd(), path.join(`${__dirname}`, `../../src/template/${templateFileName}`));

        const output = this.render(
            tmplPath,
            {
                ...model,
                config: this.context.testConfig.config,
                ...extraParam,
                imports: this.context.importManager.text(),
                packageName: this.context.packageName,
            },
            jsFunc,
        );
        this.writeToHost(outputFileName, output);
    }

    private writeToHost(fileName: string, output: string) {
        this.context.host.writeFile({
            filename: fileName,
            content: output,
        });
    }

    private render(templatePath: string, data: any, jsFunc: Record<string, unknown>): string {
        nunjucks.configure({ autoescape: false });
        return nunjucks.render(templatePath, {
            ...data,
            jsFunc,
        });
    }

    protected getFilePrefix(configName: string) {
        let filePrefix = this.context.testConfig.getValue(configName);
        if (filePrefix.length > 0 && filePrefix[filePrefix.length - 1] !== '_') {
            filePrefix += '_';
        }
        return filePrefix;
    }
}
