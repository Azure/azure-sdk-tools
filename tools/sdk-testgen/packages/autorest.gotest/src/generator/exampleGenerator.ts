/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license output.pushrmation.
 *--------------------------------------------------------------------------------------------*/

import { BaseCodeGenerator } from './baseGenerator';
import { Config } from '../common/constant';
import { ExampleModel, MockTestDefinitionModel } from '@autorest/testmodeler/dist/src/core/model';
import { GoExampleModel } from '../common/model';
import { MockTestDataRender } from './mockTestGenerator';
import { getAPIParametersSig, getClientParametersSig } from '../util/codegenBridge';

export class ExampleDataRender extends MockTestDataRender {
    protected fillExampleOutput(example: GoExampleModel) {
        const op = example.operation;
        example.methodParametersPlaceholderOutput = this.toParametersOutput(getAPIParametersSig(op), example.methodParameters);
        example.clientParametersPlaceholderOutput = this.toParametersOutput(getClientParametersSig(example.operationGroup), example.clientParameters);
    }
}

export class ExampleCodeGenerator extends BaseCodeGenerator {
    public generateCode(extraParam: Record<string, unknown> = {}): void {
        for (const [_, exampleGroups] of Object.entries(MockTestDefinitionModel.groupByOperationGroup(this.context.codeModel.testModel.mockTest.exampleGroups))) {
            let exampleModel: ExampleModel = null;
            for (const exampleGroup of exampleGroups) {
                if (exampleGroup.examples.length > 0) {
                    exampleModel = exampleGroup.examples[0];
                    break;
                }
            }
            if (exampleModel === null) {
                continue;
            }

            this.renderAndWrite(
                { exampleGroups: exampleGroups },
                'exampleTest.go.njk',
                `${this.getFilePrefix(Config.exampleFilePrefix)}example_${exampleModel.operationGroup.language.go.name.toLowerCase()}_client_test.go`,
                extraParam,
            );
        }
    }
}
