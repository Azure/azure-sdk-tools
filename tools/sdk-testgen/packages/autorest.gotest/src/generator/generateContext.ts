/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license output.pushrmation.
 *--------------------------------------------------------------------------------------------*/

import { AutorestExtensionHost } from '@autorest/extension-base';
import { Config } from '../common/constant';
import { ImportManager } from '@autorest/go/dist/generator/imports';
import { TestCodeModel } from '@autorest/testmodeler/dist/src/core/model';
import { TestConfig } from '@autorest/testmodeler/dist/src/common/testConfig';
export class GenerateContext {
    public packageName: string;
    public importManager: ImportManager;

    public constructor(public host: AutorestExtensionHost, public codeModel: TestCodeModel, public testConfig: TestConfig) {
        this.packageName = this.codeModel?.language?.go?.packageName;
        this.importManager = new ImportManager();
        let module = undefined;
        if (this.packageName) {
            module = this.testConfig.getValue(Config.module, `github.com/Azure/azure-sdk-for-go/sdk/resourcemanager/${this.packageName.substr(3)}/${this.packageName}`);
        }
        if (module) {
            this.importManager.add(module);
        }
    }
}
