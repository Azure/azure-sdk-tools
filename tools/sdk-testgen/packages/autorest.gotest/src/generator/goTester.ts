/*---------------------------------------------------------------------------------------------
 *  Copyright (c) Microsoft Corporation. All rights reserved.
 *  Licensed under the MIT License. See License.txt in the project root for license output.pushrmation.
 *--------------------------------------------------------------------------------------------*/

import * as _ from 'lodash';
import { AutorestExtensionHost } from '@autorest/extension-base';
import { Config, configDefaults } from '../common/constant';
import { ExampleCodeGenerator, ExampleDataRender } from './exampleGenerator';
import { GenerateContext } from './generateContext';
import { Helper } from '@autorest/testmodeler/dist/src/util/helper';
import { MockTestCodeGenerator, MockTestDataRender } from './mockTestGenerator';
import { SampleCodeGenerator, SampleDataRender } from './sampleGenerator';
import { ScenarioTestCodeGenerator, ScenarioTestDataRender } from './scenarioTestGenerator';
import { TestCodeModeler } from '@autorest/testmodeler/dist/src/core/model';
import { TestConfig } from '@autorest/testmodeler/dist/src/common/testConfig';

export async function processRequest(host: AutorestExtensionHost): Promise<void> {
    const session = await TestCodeModeler.getSessionFromHost(host);
    const config = new TestConfig(await session.getValue(''), configDefaults);
    if (config.getValue(Config.exportCodemodel)) {
        Helper.addCodeModelDump(session, 'go-tester-pre.yaml');
    }

    const context = new GenerateContext(host, session.model, config);
    const mockTestDataRender = new MockTestDataRender(context);
    mockTestDataRender.renderData();

    const extraParam = {
        copyright: await Helper.getCopyright(session),
        sendExampleId: config.getValue(Config.sendExampleId),
        verifyResponse: config.getValue(Config.verifyResponse),
    };
    if (config.getValue(Config.generateMockTest)) {
        const mockTestCodeGenerator = new MockTestCodeGenerator(context);
        mockTestCodeGenerator.generateCode(extraParam);
    }
    if (config.getValue(Config.generateSdkExample)) {
        const exampleDataRender = new ExampleDataRender(context);
        exampleDataRender.renderData();
        const exampleCodeGenerator = new ExampleCodeGenerator(context);
        exampleCodeGenerator.generateCode(extraParam);
    }
    if (config.getValue(Config.generateScenarioTest)) {
        const scenarioTestDataRender = new ScenarioTestDataRender(context);
        scenarioTestDataRender.renderData();
        const scenarioTestCodeGenerator = new ScenarioTestCodeGenerator(context);
        scenarioTestCodeGenerator.generateCode(extraParam);
    }
    if (config.getValue(Config.generateSdkSample)) {
        const sampleDataRender = new SampleDataRender(context);
        sampleDataRender.renderData();
        const sampleCodeGenerator = new SampleCodeGenerator(context);
        sampleCodeGenerator.generateCode(extraParam);
    }
    await Helper.outputToModelerfour(host, session);
    if (config.getValue(Config.exportCodemodel)) {
        Helper.addCodeModelDump(session, 'go-tester.yaml');
    }
    Helper.dump(host);
}
