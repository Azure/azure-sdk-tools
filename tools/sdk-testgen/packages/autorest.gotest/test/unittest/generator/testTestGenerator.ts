import { MockTool } from '../../tools';
import { TestCodeModel, TestCodeModeler } from '@autorest/testmodeler/dist/src/core/model';
import { TestConfig } from '@autorest/testmodeler/dist/src/common/testConfig';
import { TestGenerator } from '../../../src/generator/testGenerator';

class FakeGenerator extends TestGenerator {
    public getMockTestFilename() {
        return `mock_test.go`;
    }
    public genRenderDataForMock() {
        // Fake genRenderData
    }
    public genRenderDataForDefinition(_) {
        // Fake genRenderDataForDefinition
    }
    public getScenarioTestFilename(_): string {
        return 'Fake getScenarioTestFilename';
    }
}

describe('TestGenerator functions', () => {
    afterEach(() => {
        jest.restoreAllMocks();
    });

    it('generateMockTest', async () => {
        const codeModel = MockTool.createCodeModel();
        const testCodeModel = TestCodeModeler.createInstance(codeModel as TestCodeModel, {
            testmodeler: {
                'split-parents-value': true,
            },
        });
        testCodeModel.genMockTests();

        const mockWriteFile = jest.fn();
        const generator = new FakeGenerator(
            {
                // eslint-disable-next-line @typescript-eslint/naming-convention
                WriteFile: mockWriteFile,
            } as any,
            testCodeModel.codeModel,
            new TestConfig({}),
        );
        generator.genRenderData = jest.fn();
        const spyRender = jest.spyOn(FakeGenerator.prototype, 'render').mockImplementation();
        generator.genRenderData();
        await generator.generateMockTest('fake.njk');

        const generationTimes = 1;
        expect(generator.genRenderData).toBeCalledTimes(generationTimes);
        expect(spyRender).toBeCalledTimes(generationTimes);
        expect(mockWriteFile).toBeCalledTimes(generationTimes);
    });

    it('parseOavVariable', async () => {
        const generator = new FakeGenerator(undefined, undefined, new TestConfig({}));

        expect(generator.parseOavVariable('$(abc)', { abc: 'fake' })).toMatchSnapshot();
        expect(generator.parseOavVariable('$(abc)', { ac: 'fake' })).toMatchSnapshot();
        expect(generator.parseOavVariable('$(abc) with tail txt', { abc: 'fake' })).toMatchSnapshot();
        expect(generator.parseOavVariable('head texted $(abc)', { abc: 'fake' })).toMatchSnapshot();
        expect(
            generator.parseOavVariable('multiple $(abc) variables $(dbe) ... $(abc).', {
                abc: 'fake',
                dbe: 'fake2',
            }),
        ).toMatchSnapshot();
    });
});
