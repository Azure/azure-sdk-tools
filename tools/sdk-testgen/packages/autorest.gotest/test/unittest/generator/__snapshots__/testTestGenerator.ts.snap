// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TestGenerator functions parseOavVariable 1`] = `
Array [
  "abc",
]
`;

exports[`TestGenerator functions parseOavVariable 2`] = `
Array [
  "\\"$(abc)\\"",
]
`;

exports[`TestGenerator functions parseOavVariable 3`] = `
Array [
  "abc",
  "\\" with tail txt\\"",
]
`;

exports[`TestGenerator functions parseOavVariable 4`] = `
Array [
  "\\"head texted \\"",
  "abc",
]
`;

exports[`TestGenerator functions parseOavVariable 5`] = `
Array [
  "\\"multiple \\"",
  "abc",
  "\\" variables \\"",
  "dbe",
  "\\" ... \\"",
  "abc",
  "\\".\\"",
]
`;
